### ISSUE TITLE
[weight changed] export model to onnx

### ISSUE THREAD

**Issue by raisinglc** [ultralytics/yolov5#12031]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

PyTorch
1.8.1

I followed the instruction to train yolov5 with an additional tails. All original modules are frozen.  After I export trained ckpt to ONNX, the weight of the first conv layer changed. Although BN would be merged into Conv, i don’t think a frozen conv and bn would result in a different result.

Further, I modified the smart_optimizer to filter out all parameters with requires_grad=False. 
During training, I print the weight of the first conv layer. There is no change after every step, so i think the parameters frozen correctly. However, the weight of export onnx model was still different.

If I export model at very beginning of training, the export model was the same with the pretrained model.

I’am so confusing :(

**Comment by raisinglc** [comment]
The problem is the “running” parameters of bn , so bn should be set to eval if we want to freeze a model with bn.

“””
def set_bn_eval(m):
    classname = m.__class__.__name__
    if classname.find(“Batchnorm”) != -1
        m.eval()

model.apply(set_bn_eval)
“””

**Comment by glenn-jocher** [comment]
@raisinglc yes, you are right! When freezing a model with batch normalization (BN) layers, it is important to set the BN layers to evaluation mode using `model.eval()`.

You can use the provided code snippet to set all BN layers to evaluation mode:
```python
def set_bn_eval(m):
    classname = m.__class__.__name__
    if "Batchnorm" in classname:
        m.eval()

model.apply(set_bn_eval)
```
This code will recursively iterate through all the modules in `model` and set any BN layer it encounters to evaluation mode.

By setting the BN layers to evaluation mode, you ensure that the running statistics (mean and variance) of the BN layers are not updated during inference or export.

I hope this clarifies the issue for you. Let me know if you have any further questions!
